File: cherokee/table.c
/* -*- Mode: C; tab-width: 8; indent-tabs-mode: t; c-basic-offset: 8 -*- */

/* Cherokee
 *
 * Authors:
 *      Alvaro Lopez Ortega <alvaro@alobbs.com>
 *
 * Copyright (C) 2001-2011 Alvaro Lopez Ortega
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of version 2 of the GNU General Public
 * License as published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
 * 02110-1301, USA.
 */

#include "common-internal.h"
#include "table.h"

ret_t
cherokee_table_init (cherokee_table_t *tab)
{
	ret_t ret;

	ret = cherokee_avl_init (&tab->avl);
	if (unlikely (ret != ret_ok)) return ret;

	tab->case_sensitive = false;
	return ret_ok;
}

ret_t
cherokee_table_init_case (cherokee_table_t *tab)
{
	ret_t ret;

	ret = cherokee_avl_init (&tab->avl);
	if (unlikely (ret != ret_ok)) return ret;

	tab->case_sensitive = true;
	return ret_ok;
}


ret_t
cherokee_table_free2 (cherokee_table_t *tab, cherokee_table_free_item_t func)
{
	cherokee_table_mrproper2 (tab, func);
	free(tab);
	return ret_ok;
}


ret_t
cherokee_table_clean (cherokee_table_t *tab)
{
	return cherokee_table_clean2 (tab, NULL);
}

static int
free_entry (const char *key, void *value, void *param)
{
	cherokee_table_free_item_t func = param;
	func (value);
	return 0;
}


ret_t
cherokee_table_clean2 (cherokee_table_t *tab, cherokee_table_free_item_t func)
{
	return cherokee_avl_while (&tab->avl, free_entry, func, NULL, NULL);
}


ret_t
cherokee_table_mrproper (cherokee_table_t *tab)
{
	return cherokee_table_clean (tab);
}


ret_t
cherokee_table_mrproper2 (cherokee_table_t *tab, cherokee_table_free_item_t func)
{
	return cherokee_table_clean2 (tab, func);
}


CHEROKEE_ADD_FUNC_NEW(table);
CHEROKEE_ADD_FUNC_FREE(table);


ret_t
cherokee_table_add (cherokee_table_t *tab, char *key, void *value)
{
	cherokee_buffer_t tmp;

	cherokee_buffer_fake (&tmp, key, strlen(key));
	return cherokee_avl_add (&tab->avl, &tmp, value);
}


ret_t
cherokee_table_get (cherokee_table_t *tab, char *key, void **val)
{
	cherokee_buffer_t tmp;

	cherokee_buffer_fake (&tmp, key, strlen(key));
	return cherokee_avl_get (&tab->avl, &tmp, val);
}


ret_t
cherokee_table_del (cherokee_table_t *tab, char *key, void **val)
{
	cherokee_buffer_t tmp;

	cherokee_buffer_fake (&tmp, key, strlen(key));
	return cherokee_avl_get (&tab->avl, &tmp, val);
}


ret_t
cherokee_table_len (cherokee_table_t *tab, size_t *len)
{
	return cherokee_avl_len (&tab->avl, len);
}


static ret_t
each_func_avl_to_table (cherokee_buffer_t *key, void *value, void *param)
{
	// To be removed
	cherokee_table_foreach_func_t func = param;
	func (key->buf, value);
	return ret_ok;
}

ret_t
cherokee_table_foreach (cherokee_table_t *tab, cherokee_table_foreach_func_t func)
{
	return cherokee_avl_while (&tab->avl, each_func_avl_to_table, func, NULL, NULL);
}


ret_t
while_func_avl_to_table (cherokee_buffer_t *key, void *value, void *param)
{
	// To be removed
	int                          re;
	void                       **params   = param;
	cherokee_table_while_func_t  func     = params[0];
	void                        *data     = params[1];

	re = func (key->buf, value, data);
	if (re != 0) return ret_not_found;

	return ret_ok;
}


ret_t
cherokee_table_while (cherokee_table_t *tab, cherokee_table_while_func_t func, void *param, char **key, void **value)
{
	ret_t              ret;
	cherokee_buffer_t *key_buf  = NULL;
	void              *params[] = { func, param };

	ret = cherokee_avl_while (&tab->avl, while_func_avl_to_table, (void *)params, &key_buf, value);
	if (key && key_buf)
		*key = key_buf->buf;

	return ret;
}

File: Table.py
# CTK: Cherokee Toolkit
#
# Authors:
#      Alvaro Lopez Ortega <alvaro@alobbs.com>
#
# Copyright (C) 2009 Alvaro Lopez Ortega
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of version 2 of the GNU General Public
# License as published by the Free Software Foundation.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
# 02110-1301, USA.
#

from Widget import Widget, RenderResponse
from Container import Container


class TableField (Container):
    def __init__ (self, widget=None):
        Container.__init__ (self)
        self._props = {}
        self._tag   = 'td'

        if widget:
            self += widget

    def Render (self):
        # Build tag props
        props = ''
        for p in self._props:
            value = self._props[p]
            if value:
                props += ' %s="%s"' %(p, value)
            else:
                props += ' %s' %(p)

        # Render content
        render = Container.Render(self)

        # Output
        html  = '<%s%s>' % (self._tag, props)
        html += render.html
        html += '</%s>' % (self._tag)

        r = RenderResponse (html, render.js, render.headers)
        return r

    def __setitem__ (self, prop, value):
        assert type(prop)  == str
        assert type(value) == str

        self._props[prop] = value


class Table (Widget):
    def __init__ (self, **kwargs):
        Widget.__init__ (self)
        self.kwargs   = kwargs
        self.last_col = 0
        self.last_row = 0
        self.rows     = []

    def __add_row (self):
        self.rows.append ([None] * self.last_col)
        self.last_row += 1

    def __add_col (self):
        for row in self.rows:
            row.append (None)
        self.last_col += 1

    def __grow_if_needed (self, row, col):
        if col < 1:
            raise IndexError, "Column number out of bounds"
        if row < 1:
            raise IndexError, "Row number out of bounds"

        while row > self.last_row:
            self.__add_row()

        while col > self.last_col:
            self.__add_col()

    def __setitem_single (self, pos, widget):
        assert isinstance(widget, Widget)
        assert type(pos) == tuple

        # Check table bounds
        row,col = pos
        self.__grow_if_needed (row, col)

        # The field object
        field = TableField (widget)

        # Set the element
        table_row = self.rows[row-1]
        table_row[col-1] = field

    def __setitem_list (self, pos, wid_list):
        row,col = pos
        for widget in wid_list:
            self.__setitem_single ((row,col), widget)
            col += 1

    def __setitem__ (self, pos, item):
        if type(item) == list:
            return self.__setitem_list (pos, item)
        else:
            return self.__setitem_single (pos, item)

    def __getitem__ (self, pos):
        row,col = pos

        if col > self.last_col:
            raise IndexError, "Column number out of bounds"
        if row > self.last_row:
            raise IndexError, "Row number out of bounds"

        return self.rows[row-1][col-1]

    def Render (self):
        props = self.kwargs.get('props')
        if props:
            html = '<table %s>' %(props)
        else:
            html = '<table>'

        render = RenderResponse (html)

        for row in self.rows:
            html += '<tr>'
            for field in row:
                if field:
                    assert isinstance(field, TableField)
                    r = field.Render()
                    render += r
            render.html += '</tr>'

        render.html += '</table>'
        return render


class TableFixed (Table):
    def __init__ (self, rows, cols):
        Table.__init__ (self)
        self._fixed_rows = rows
        self._fixed_cols = cols

    def __setitem__ (self, pos, item):
        row,col = pos

        if row > self._fixed_rows:
            raise IndexError, "Row number out of bounds"

        if col > self._fixed_cols:
            raise IndexError, "Column number out of bounds"

        return Table.__setitem__ (self, pos, item)
